File: LeetCode Practice Problems/155. Min Stack.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    stack<int> s;
    MinStack() {
    }
    
    void push(int x) {
        s.push(x);
    }
    
    void pop() {
        s.pop();
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        int mini = INT_MAX;
        stack<int> st;
        while(!s.empty())
        {
            mini = min(mini , s.top());
            st.push(s.top());
            s.pop();
        }
        
        while(!st.empty())
        {
            s.push(st.top());
            st.pop();
        }
        return mini;
    }
};

File: LeetCode/LeetCode Practice Problems/155-Min Stack.java
class MinStack {
    static class Node {
        int val,mn;
        Node next;
        Node(int val,int mn) {
            this.val = val;
            this.mn = mn;
            this.next = null;
        }
    }
    Node head;  
    public MinStack() {
        head = null;
    }
    public void push(int val) {
        if(head == null) {
            head = new Node(val,val);
        }
        else {
            int curMin = head.mn;
            Node node = new Node(val , Math.min(val , curMin));
            node.next = head;
            head = node;
        }
    }
    public void pop() {
        head = head.next;
    }
    public int top() {
        return head.val;
    }
    public int getMin() {
        return head.mn;
    }
}
